todos/src/components/Navbar.js

```diff
@@ -1,17 +1,59 @@
 import logo from "icons/logo.svg";
 import search from "icons/searchIcon.svg";
-import "styles/navbar.css"
+import "styles/navbar.css";
+import { useSelector, useDispatch } from "react-redux";
+import { setIconVisibility, setSearchValue, setLoadingState } from "actions";
+import { useCallback } from "react";
+import spinner from "icons/spinner.svg";
+import "styles/loading.css"
 
 const Navbar = () => {
+  const dispatch = useDispatch();
+  const searchIconState = useSelector((state) => state.searchReducer.iconState);
+  const handleSearchIconClick = () => {
+    dispatch(setIconVisibility(!searchIconState));
+  };
+
+  const handleSearchInput = (event) => {
+    dispatch(setSearchValue(event.target.value));
+  };
+
+  const loadingState = useSelector(
+    (state) => state.laodingReducer.loadingState
+  );
+
+  const debounce = (func) => {
+    let timer;
+    return function (...args) {
+      const context = this;
+      dispatch(setLoadingState(true));
+      if (timer) clearTimeout(timer);
+      timer = setTimeout(() => {
+        timer = null;
+        func.apply(context, args);
+        dispatch(setLoadingState(false));
+      }, 1000);
+      
+    };
+  };
```

Comment:
Let's generalize the debounce function and move it to `utils.js` file for modularity.
```suggestion
  // utils.js
  export const debounce = (func, delay) => {
    let timer;
    return function (...args) {
      const context = this;
      if (timer) clearTimeout(timer);
      timer = setTimeout(() => {
        timer = null;
        func.apply(context, args);
      }, delay);
    };
  };
```